src/lib/src/x86_calcbonds.cpp

```diff
@@ -30,37 +33,85 @@ constexpr std::size_t kStreamingThreshold = 16 * 1024 * 1024;
 template <bool streaming_store, typename VectorT, typename BoxT>
 void CalcBondsInner(const VectorToScalarT<VectorT> *coords0,
                     const VectorToScalarT<VectorT> *coords1,
-                    const VectorToScalarT<VectorT> *box,
-                    std::size_t n,
+                    const VectorToScalarT<VectorT> *box, std::size_t n,
                     VectorToScalarT<VectorT> *out) {
   auto vecbox = BoxT(box);
 
-  size_t i = 0;
+  std::size_t i = 0;
   if (n % ValuesPerPack<VectorT>) {
     auto c0 = VectorTriple<VectorT>(coords0);
     auto c1 = VectorTriple<VectorT>(coords1);
 
     VectorT result = NewDistance3dWithBoundary(c0, c1, vecbox);
-
+    // TODO constexpr if with CXX17 support
     if (streaming_store) {
       genericstream(out, result);
     } else {
       genericstore(out, result);
     }
     i += n % ValuesPerPack<VectorT>;
   }
-  for (;i<n;i+=ValuesPerPack<VectorT>) {
-    auto c0 = VectorTriple<VectorT>(&coords0[3*i]);
-    auto c1 = VectorTriple<VectorT>(&coords1[3*i]);
+  for (; i < n; i += ValuesPerPack<VectorT>) {
+    auto c0 = VectorTriple<VectorT>(&coords0[3 * i]);
+    auto c1 = VectorTriple<VectorT>(&coords1[3 * i]);
 
     VectorT result = NewDistance3dWithBoundary(c0, c1, vecbox);
+    // TODO constexpr if with CXX17 support
     if (streaming_store) {
       genericstream(&out[i], result);
     } else {
       genericstore(&out[i], result);
     }
   }
+  // TODO constexpr if with CXX17 support
+  if (streaming_store) {
+    _mm_mfence();
+  }
+}
+
+template <bool streaming_store, typename VectorT, typename BoxT>
+void CalcBondsIdxInner(const VectorToScalarT<VectorT> *coords,
+                       const std::size_t *idxs,
+                       const VectorToScalarT<VectorT> *box, std::size_t n,
+                       VectorToScalarT<VectorT> *out) {
+  auto vecbox = BoxT(box);
+  std::size_t i = 0;
+  // indicies of the bonds
+  const std::size_t *b_i = idxs;
+  const std::size_t *b_j = idxs + 1;
+
+  if (n % ValuesPerPack<VectorT>) {
+    // WARNING BROKEN USE OF some big buffer
+    auto c0 = VectorTriple<VectorT>(coords, coords + 1000000, b_i, 2);
+    auto c1 = VectorTriple<VectorT>(coords, coords + 1000000, b_j, 2);
+    VectorT result = NewDistance3dWithBoundary(c0, c1, vecbox);
+    // TODO constexpr if with CXX17 support
+    if (streaming_store) {
+      genericstream(out, result);
+    } else {
+      genericstore(out, result);
+    }
+    i += n % ValuesPerPack<VectorT>;
+  }
+  for (; i < n; i += ValuesPerPack<VectorT>) {
+    // access with stride of 2
+    // WARNING Abuse of big buffer  until loader can be fixed.
+    auto c0 = VectorTriple<VectorT>(coords, coords + 1000000, b_i, 2);
```

Comment:
this needs to get split into two lines:
```c++
t = VectorTriple()
t.load<2>(coords, b_i)
```

Reply:
Why default construct then load if we can just one shot it ?

Reply:
I see why this is better now. I have fixed it.